[ICE]: `can't get the span of an arbitrary parsed attribute`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_hir/src/hir.rs:1390:18: 'can't get the span of an arbitrary parsed attribute: Parsed(Doc(DocAttribute { aliases: {":(": /tmp/im/a.rs:4:19: 4:23 (#0)}, hidden: None, inline: [], cfg: [], auto_cfg: [], auto_cfg_change: [], fake_variadic: None, keyword: None, attribute: None, masked: None, notable_trait: None, search_unbox: None, html_favicon_url: None, html_logo_url: None, html_playground_url: None, html_root_url: None, html_no_source: None, issue_tracker_base_url: None, rust_logo: None, test_attrs: [], no_crate_inject: None }))'', 'thread 'rustc' ($TID) panicked at compiler/rustc_hir/src/hir.rs:1390:18: 'can't get the span of an arbitrary parsed attribute: Parsed(Doc(DocAttribute { aliases: {":(": /tmp/im/a.rs:4:19: 4:23 (#0)}, hidden: None, inline: [], cfg: [], auto_cfg: [], auto_cfg_change: [], fake_variadic: None, keyword: None, attribute: None, masked: None, notable_trait: None, search_unbox: None, html_favicon_url: None, html_logo_url: None, html_playground_url: None, html_root_url: None, html_no_source: None, issue_tracker_base_url: None, rust_logo: None, test_attrs: [], no_crate_inject: None }))''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
fn test()
where
    #[doc(alias = ":(")]
    ():,
{
}

````

original:
````rust
#![feature(where_clause_attrs)]
fn test()
where
    #[doc(alias = ":(")]
    (): {}

````

Version information
````
rustc 1.97.0-nightly (68ffae46b 2026-04-25)
binary: rustc
commit-hash: 68ffae46b581747074413e4242ab0c6ecc4db4a9
commit-date: 2026-04-25
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/68ffae46b581747074413e4242ab0c6ecc4db4a9/compiler/rustc_hir/src/hir.rs#L1384-L1396

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: attributes in `where` clause are unstable
 --> /tmp/icemaker_global_tempdir.ayTSYC5T6eHJ/rustc_testrunner_tmpdir_reporting.vpANhFq5WUiX/mvce.rs:3:5
  |
3 |     #[doc(alias = ":(")]
  |     ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #115590 <https://github.com/rust-lang/rust/issues/115590> for more information
  = help: add `#![feature(where_clause_attrs)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-25; consider upgrading it if it is out of date


thread 'rustc' (366159) panicked at /rustc-dev/68ffae46b581747074413e4242ab0c6ecc4db4a9/compiler/rustc_hir/src/hir.rs:1390:18:
can't get the span of an arbitrary parsed attribute: Parsed(Doc(DocAttribute { aliases: {":(": /tmp/icemaker_global_tempdir.ayTSYC5T6eHJ/rustc_testrunner_tmpdir_reporting.vpANhFq5WUiX/mvce.rs:3:19: 3:23 (#0)}, hidden: None, inline: [], cfg: [], auto_cfg: [], auto_cfg_change: [], fake_variadic: None, keyword: None, attribute: None, masked: None, notable_trait: None, search_unbox: None, html_favicon_url: None, html_logo_url: None, html_playground_url: None, html_root_url: None, html_no_source: None, issue_tracker_base_url: None, rust_logo: None, test_attrs: [], no_crate_inject: None }))
stack backtrace:
   0:     0x7f2cbdf62edb - <<std[74a5d452ef4468e2]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[af918ca625322054]::fmt::Display>::fmt
   1:     0x7f2cbe62d948 - core[af918ca625322054]::fmt::write
   2:     0x7f2cbdf79b16 - <std[74a5d452ef4468e2]::sys::stdio::unix::Stderr as std[74a5d452ef4468e2]::io::Write>::write_fmt
   3:     0x7f2cbdf39158 - std[74a5d452ef4468e2]::panicking::default_hook::{closure#0}
   4:     0x7f2cbdf56543 - std[74a5d452ef4468e2]::panicking::default_hook
   5:     0x7f2cbcf2cb9c - std[74a5d452ef4468e2]::panicking::update_hook::<alloc[3c1e8b5c2310114c]::boxed::Box<rustc_driver_impl[65db170643483585]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f2cbdf56822 - std[74a5d452ef4468e2]::panicking::panic_with_hook
   7:     0x7f2cbdf39218 - std[74a5d452ef4468e2]::panicking::panic_handler::{closure#0}
   8:     0x7f2cbdf2d779 - std[74a5d452ef4468e2]::sys::backtrace::__rust_end_short_backtrace::<std[74a5d452ef4468e2]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f2cbdf3ac7d - __rustc[4d529550cd7027cf]::rust_begin_unwind
  10:     0x7f2cbab439fc - core[af918ca625322054]::panicking::panic_fmt
  11:     0x7f2cbca1d04d - <core[af918ca625322054]::iter::adapters::map::Map<core[af918ca625322054]::iter::adapters::filter::Filter<core[af918ca625322054]::iter::adapters::chain::Chain<core[af918ca625322054]::slice::iter::Iter<rustc_hir[3704b234a2d4cda3]::hir::Attribute>, core[af918ca625322054]::slice::iter::Iter<rustc_hir[3704b234a2d4cda3]::hir::Attribute>>, <rustc_attr_parsing[ee812df563a5c05e]::interface::AttributeParser>::check_invalid_where_predicate_attrs<core[af918ca625322054]::iter::adapters::chain::Chain<core[af918ca625322054]::slice::iter::Iter<rustc_hir[3704b234a2d4cda3]::hir::Attribute>, core[af918ca625322054]::slice::iter::Iter<rustc_hir[3704b234a2d4cda3]::hir::Attribute>>>::{closure#0}>, <rustc_attr_parsing[ee812df563a5c05e]::interface::AttributeParser>::check_invalid_where_predicate_attrs<core[af918ca625322054]::iter::adapters::chain::Chain<core[af918ca625322054]::slice::iter::Iter<rustc_hir[3704b234a2d4cda3]::hir::Attribute>, core[af918ca625322054]::slice::iter::Iter<rustc_hir[3704b234a2d4cda3]::hir::Attribute>>>::{closure#1}> as core[af918ca625322054]::iter::traits::iterator::Iterator>::next
  12:     0x7f2cbecf13dc - <rustc_ast_lowering[65b0b6d9972c4088]::LoweringContext>::lower_attrs_vec
  13:     0x7f2cbecec1e3 - <rustc_ast_lowering[65b0b6d9972c4088]::LoweringContext>::lower_attrs_with_extra
  14:     0x7f2cbecc743e - <rustc_ast_lowering[65b0b6d9972c4088]::LoweringContext>::lower_where_predicate
  15:     0x7f2cbece4151 - <rustc_ast_lowering[65b0b6d9972c4088]::LoweringContext>::lower_item_kind
  16:     0x7f2cbecfe286 - <rustc_ast_lowering[65b0b6d9972c4088]::item::ItemLowerer>::lower_node
  17:     0x7f2cbecfc890 - rustc_ast_lowering[65b0b6d9972c4088]::lower_to_hir
  18:     0x7f2cbf76726c - rustc_query_impl[2514449be8add55a]::query_impl::hir_crate::invoke_provider_fn::__rust_begin_short_backtrace
  19:     0x7f2cbf764c19 - rustc_query_impl[2514449be8add55a]::execution::try_execute_query::<rustc_middle[11ba238a743b416b]::query::caches::SingleCache<rustc_middle[11ba238a743b416b]::query::erase::ErasedData<[u8; 8usize]>>, false>
  20:     0x7f2cbf51abbf - rustc_query_impl[2514449be8add55a]::query_impl::hir_crate::execute_query_non_incr::__rust_end_short_backtrace
  21:     0x7f2cbf51ac1c - rustc_middle[11ba238a743b416b]::hir::map::hir_crate_items
  22:     0x7f2cbf767fa9 - rustc_query_impl[2514449be8add55a]::query_impl::hir_crate_items::invoke_provider_fn::__rust_begin_short_backtrace
  23:     0x7f2cbf764c19 - rustc_query_impl[2514449be8add55a]::execution::try_execute_query::<rustc_middle[11ba238a743b416b]::query::caches::SingleCache<rustc_middle[11ba238a743b416b]::query::erase::ErasedData<[u8; 8usize]>>, false>
  24:     0x7f2cbf5076b9 - rustc_query_impl[2514449be8add55a]::query_impl::hir_crate_items::execute_query_non_incr::__rust_end_short_backtrace
  25:     0x7f2cbf50801e - rustc_interface[78b66ebcbd67330e]::passes::analysis
  26:     0x7f2cbf7637c9 - rustc_query_impl[2514449be8add55a]::execution::try_execute_query::<rustc_middle[11ba238a743b416b]::query::caches::SingleCache<rustc_middle[11ba238a743b416b]::query::erase::ErasedData<[u8; 0usize]>>, false>
  27:     0x7f2cbf763429 - rustc_query_impl[2514449be8add55a]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  28:     0x7f2cbf72e779 - rustc_interface[78b66ebcbd67330e]::interface::run_compiler::<(), rustc_driver_impl[65db170643483585]::run_compiler::{closure#0}>::{closure#1}
  29:     0x7f2cbf754584 - std[74a5d452ef4468e2]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[78b66ebcbd67330e]::util::run_in_thread_with_globals<rustc_interface[78b66ebcbd67330e]::util::run_in_thread_pool_with_globals<rustc_interface[78b66ebcbd67330e]::interface::run_compiler<(), rustc_driver_impl[65db170643483585]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  30:     0x7f2cbf754c6d - <std[74a5d452ef4468e2]::thread::lifecycle::spawn_unchecked<rustc_interface[78b66ebcbd67330e]::util::run_in_thread_with_globals<rustc_interface[78b66ebcbd67330e]::util::run_in_thread_pool_with_globals<rustc_interface[78b66ebcbd67330e]::interface::run_compiler<(), rustc_driver_impl[65db170643483585]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[af918ca625322054]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  31:     0x7f2cbf755aec - <std[74a5d452ef4468e2]::sys::thread::unix::Thread>::new::thread_start
  32:     0x7f2cb92a597a - <unknown>
  33:     0x7f2cb93292bc - <unknown>
  34:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.97.0-nightly (68ffae46b 2026-04-25) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [hir_crate] getting the crate HIR
#1 [hir_crate_items] getting HIR crate items
#2 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
#0 [hir_crate] getting the crate HIR
#1 [hir_crate_items] getting HIR crate items
#2 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-where_clause_attrs 


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_attr_parsing/src/attributes/doc.rs
+++ b/compiler/rustc_attr_parsing/src/attributes/doc.rs
@@ -701,6 +701,9 @@ impl DocParser {
                 for i in items.mixed() {
                     match i {
                         MetaItemOrLitParser::MetaItemParser(mip) => {
+                            if self.nb_doc_attrs == 0 {
+                                self.attribute.first_span = cx.attr_span;
+                            }
                             self.nb_doc_attrs += 1;
                             self.parse_single_doc_attr_item(cx, mip);
                         }

--- a/compiler/rustc_attr_parsing/src/target_checking.rs
+++ b/compiler/rustc_attr_parsing/src/target_checking.rs
@@ -292,15 +292,16 @@ impl<'sess, S: Stage> AttributeParser<'sess, S> {
         // in where clauses. After that, this function would become useless.
         let spans = attrs
             .into_iter()
-            // FIXME: We shouldn't need to special-case `doc`!
-            .filter(|attr| {
-                matches!(
-                    attr,
-                    Attribute::Parsed(AttributeKind::DocComment { .. } | AttributeKind::Doc(_))
-                        | Attribute::Unparsed(_)
-                )
+            .filter_map(|attr| {
+                match attr {
+                    Attribute::Parsed(AttributeKind::DocComment { span, .. }) => Some(*span),
+                    // FIXME: We shouldn't need to special-case `doc`!
+                    Attribute::Parsed(AttributeKind::Doc(attr)) => Some(attr.first_span),
+                    // Checked during attribute parsing target checking
+                    Attribute::Parsed(_) => None,
+                    Attribute::Unparsed(attr) => Some(attr.span),
+                }
             })
-            .map(|attr| attr.span())
             .collect::<Vec<_>>();
         if !spans.is_empty() {
             self.dcx()

--- a/compiler/rustc_hir/src/attrs/data_structures.rs
+++ b/compiler/rustc_hir/src/attrs/data_structures.rs
@@ -544,6 +544,8 @@ pub struct CfgHideShow {
 
 #[derive(Clone, Debug, Default, HashStable_Generic, Decodable, PrintAttribute)]
 pub struct DocAttribute {
+    pub first_span: Span,
+
     pub aliases: FxIndexMap<Symbol, Span>,
     pub hidden: Option<Span>,
     // Because we need to emit the error if there is more than one `inline` attribute on an item
@@ -581,6 +583,7 @@ pub struct DocAttribute {
 impl<E: rustc_span::SpanEncoder> rustc_serialize::Encodable<E> for DocAttribute {
     fn encode(&self, encoder: &mut E) {
         let DocAttribute {
+            first_span,
             aliases,
             hidden,
             inline,
@@ -603,6 +606,7 @@ impl<E: rustc_span::SpanEncoder> rustc_serialize::Encodable<E> for DocAttribute 
             test_attrs,
             no_crate_inject,
         } = self;
+        rustc_serialize::Encodable::<E>::encode(first_span, encoder);
         rustc_serialize::Encodable::<E>::encode(aliases, encoder);
         rustc_serialize::Encodable::<E>::encode(hidden, encoder);
 

--- a/compiler/rustc_passes/src/check_attr.rs
+++ b/compiler/rustc_passes/src/check_attr.rs
@@ -1027,6 +1027,7 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
     /// [`check_doc_inline`]: Self::check_doc_inline
     fn check_doc_attrs(&self, attr: &DocAttribute, hir_id: HirId, target: Target) {
         let DocAttribute {
+            first_span: _,
             aliases,
             // valid pretty much anywhere, not checked here?
             // FIXME: should we?

--- a/src/librustdoc/clean/mod.rs
+++ b/src/librustdoc/clean/mod.rs
@@ -2780,6 +2780,7 @@ fn add_without_unwanted_attributes<'hir>(
             hir::Attribute::Parsed(AttributeKind::Doc(box d)) => {
                 // Remove attributes from `normal` that should not be inherited by `use` re-export.
                 let DocAttribute {
+                    first_span: _,
                     aliases,
                     hidden,
                     inline,

--- a/src/librustdoc/json/conversions.rs
+++ b/src/librustdoc/json/conversions.rs
@@ -960,6 +960,7 @@ fn maybe_from_hir_attr(attr: &hir::Attribute, item_id: ItemId, tcx: TyCtxt<'_>) 
             }
 
             let DocAttribute {
+                first_span: _,
                 aliases,
                 hidden,
                 inline,

--- a/tests/ui/attributes/where-doc.rs
+++ b/tests/ui/attributes/where-doc.rs
@@ -16,6 +16,19 @@ where
 //~^ ERROR most attributes are not supported in `where` clauses
 ():,
 
+// == That the doc attributes below don't trigger the error is a bug
+#[doc()]
+():,
+
+#[doc(5)]
+():,
+
+#[doc]
+():,
+
+#[doc = 5]
+():,
+
 { }
 
 fn main() {}

--- a/tests/ui/attributes/where-doc.rs
+++ b/tests/ui/attributes/where-doc.rs
@@ -1,0 +1,21 @@
+#![feature(where_clause_attrs)]
+#![allow(invalid_doc_attributes)]
+
+fn test()
+where
+#[doc(alias = ":(")]
+//~^ ERROR most attributes are not supported in `where` clauses
+//~| ERROR `#[doc(alias = "...")]` isn't allowed on where predicate
+():,
+
+#[doc(hidden)]
+//~^ ERROR most attributes are not supported in `where` clauses
+():,
+
+#[doc = ""]
+//~^ ERROR most attributes are not supported in `where` clauses
+():,
+
+{ }
+
+fn main() {}

--- a/tests/ui/attributes/where-doc.stderr
+++ b/tests/ui/attributes/where-doc.stderr
@@ -1,0 +1,32 @@
+error: most attributes are not supported in `where` clauses
+  --> $DIR/where-doc.rs:6:1
+   |
+LL | #[doc(alias = ":(")]
+   | ^^^^^^^^^^^^^^^^^^^^
+   |
+   = help: only `#[cfg]` and `#[cfg_attr]` are supported
+
+error: most attributes are not supported in `where` clauses
+  --> $DIR/where-doc.rs:11:1
+   |
+LL | #[doc(hidden)]
+   | ^^^^^^^^^^^^^^
+   |
+   = help: only `#[cfg]` and `#[cfg_attr]` are supported
+
+error: most attributes are not supported in `where` clauses
+  --> $DIR/where-doc.rs:15:1
+   |
+LL | #[doc = ""]
+   | ^^^^^^^^^^^
+   |
+   = help: only `#[cfg]` and `#[cfg_attr]` are supported
+
+error: `#[doc(alias = "...")]` isn't allowed on where predicate
+  --> $DIR/where-doc.rs:6:15
+   |
+LL | #[doc(alias = ":(")]
+   |               ^^^^
+
+error: aborting due to 4 previous errors
+

